### PR TITLE
[Api] Prevent product more than in stock

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/AddingEligibleProductVariantToCartValidator.php [hunk 1]
```diff
@@ -85,7 +86,31 @@ public function validate($value, Constraint $constraint)
...
+
+                /** @var ProductVariantInterface */
+                $variant = $orderItem->getVariant();
+
+                $code = $variant->getCode();
+
+                if ($value->productVariantCode = $code) {
+                    ++$sameOrderItemsCounter;
+                }
+            }
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/AddingEligibleProductVariantToCartValidator.php:110]: I am wondering about the performance of this solution. In cart, we can have a really big amount of item units, and every addition to the cart will iterate over the entire cart...

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/AddingEligibleProductVariantToCartValidator.php [hunk 2]
```diff
@@ -85,7 +86,31 @@ public function validate($value, Constraint $constraint)
...
+        }
+
+        if (!$this->availabilityChecker->isStockSufficient($productVariant, $value->quantity + $sameOrderItemsCounter)) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/AddingEligibleProductVariantToCartValidator.php:113]: You can create a private method for providing a quantity of given product in the cart, sth like:
```
    private function getExistingCartItemQuantityFromCart(OrderInterface $cart, ProductVariantInterface $productVariant): int
    {
        foreach ($cart->getItems() as $existingCartItem) {
            if ($existingCartItem-> getVariant()->equals($productVariant)) {
                return $existingCartItem->getQuantity();
            }
        }

        return 0;
    }
```
and then
```suggestion
        if (!$this->availabilityChecker->isStockSufficient($productVariant, $value->quantity + $this->getExistingCartItemQuantityFromCart($cart, $productVariant))) {
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/AddingEligibleProductVariantToCartValidator.php [hunk 3]
```diff
@@ -107,4 +111,15 @@ public function validate($value, Constraint $constraint)
             );
         }
     }
+
+    private function getExistingCartItemQuantityFromCart(OrderInterface $cart, ProductVariantInterface $productVariant): int
+    {
+        foreach ($cart->getItems() as $existingCartItem) {
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/AddingEligibleProductVariantToCartValidator.php:117]: I'm wondering, maybe we can use `OrderInterface::getItemUnitsByVariant($variant)` there? After calling `->count()` method we would have the number of item units for a specific variant - which indeed is a _cart item quantity_ 💃 Maybe we would even be able to resign from the additional private function:

```php
if (!$this-&gt;availabilityChecker-&gt;isStockSufficient(
    $productVariant,
    $value->quantity + $cart->getItemUnitsByVariant($variant)->count()
)) {
```

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/AddingEligibleProductVariantToCartValidatorSpec.php [hunk 4]
```diff
@@ -131,21 +134,91 @@ function it_adds_violation_if_product_variant_is_disabled(
...
+
+        $executionContext
+            ->addViolation('sylius.product_variant.not_sufficient', ['%productVariantCode%' => 'productVariantCode'])
+            ->shouldBeCalled()
+        ;
+
+        $this->validate(
+            $command,
+            new AddingEligibleProductVariantToCart()
+        );
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/AddingEligibleProductVariantToCartValidatorSpec.php:183]: Would fit in one line 🚀

**Review by Zales0123 — APPROVED** [review]
Btw, this spec file is absurdly long 😄 We should definitely think about some validator simplification, as _maybe_ it does a little bit too much 🐎

**Comment by Zales0123** [comment]
Thanks, Tomasz! :1st_place_medal:

**Pull request merged**